### PR TITLE
feat(engine): expose per-node attribute schemas via 'probe-schema' command

### PR DESCRIPTION
**Pull request by pyshx**
# Overview

Adds a static **`probe-schema`** CLI command to the engine that exposes the **available feature attributes at each node** of a workflow, so the UI/user can know what attributes exist and target them in expressions (e.g. `attributes["myAttribute"]`). Source readers are sampled (a light, bounded read of their dataset) to discover real attributes; downstream transforms (`AttributeManager`, etc.) propagate on top. Output is JSON for a server/UI to consume.

> This branch was repurposed from an earlier static `build`/`check` *validator* (which provided no standalone user value); that surface has been removed and the reusable engine kept. See "What I haven't done".

> **Naming:** the command is `probe-schema` (not bare `schema`) — `schema-action` and `schema-workflow` already exist and emit *static* JSON Schema from definitions, so the bare name was ambiguous. `probe-schema` stresses that it actually runs the workflow's readers to probe real data.

## What I've done

- **Schema model** (`reearth-flow-types`): `AttrSchema` (ordered fields + `open` flag), `AttrType`, `AttrField` (`ty` + `Presence::{Always,Maybe}`), and a `join` lattice for multi-edge fan-in. Plus a serde **`SchemaReport`** JSON DTO (ordered `fields` array, `version`, per-node `note`).
- **Source sampler** (`runtime/runtime/src/schema_sample.rs`): runs a source reader briefly against a bounded channel and folds the first N features into a **closed, typed** `AttrSchema` (default 10; `--sample-size 0` = all). Accumulates **incrementally** — each feature is observed then dropped, so memory is `O(#attributes)` not `O(#features)`, even for unbounded sampling. Per-source failures degrade to `open` + a `note`; never panics. No processors/sinks run, no sink writes.
- **Propagation** (`schema_infer.rs`): `infer_with_sampling` seeds source nodes from samples, then propagates per-port schemas through the DAG in topological order (cycle-detected). Transfer functions for `AttributeManager`, `AttributeMapper`, `StatisticsCalculator`, `DateTimeConverter`, `FeatureFilter`; unknown processors pass through. **Dynamic output ports are honored**: an `effective_output_ports` helper augments each node's declared ports with config-derived ones (OutputRouter `routingPort`, FeatureFilter `conditions[].outputPort`, mirroring `builder_dag`), so branching nodes' routed ports appear in the report.
- **Conservative attribute presence**: inference never over-claims relative to the *conditional* runtime. `AttributeManager` Create/Convert mark a newly-produced key as `Maybe` (the runtime only inserts on successful eval; an already-present key keeps its presence with type `Unknown`); Rename downgrades the source key to `Maybe` and marks the schema `open` (it may be kept on eval error / destination collision) rather than dropping it unconditionally.
- **`probe-schema` CLI command**: `reearth-flow probe-schema --workflow <path|-> [--var K=V] [--sample-size N]` → prints `{ version, sampleSize, nodes: { id: { name, ports: { port: { open, fields[] } }, note? } } }` to stdout (logs to stderr). Supports `!include` expansion and `--var`. Node ids/ports are sorted for deterministic output.

## What I haven't done

- **Removed** the previous static `build`/`check` validator (command, `referenced_input_attributes`, `Diagnostic`/`Severity`, `AttrRef`) — it wasn't independently useful; the reusable schema engine was retained.
- **Server GraphQL endpoint** and **UI display/autocomplete** — out of scope; this is the engine foundation. The `probe-schema` JSON is their contract (separate specs to follow).
- **Typed source seeds for non-feature sources** (DB catalog) and **Rhai value-type inference** — expression-derived values are `Unknown` for now (Rhai is being replaced). Sampling bounds *features processed*, not always *bytes read* (whole-file readers still read the file).

## How I tested

- Unit tests: `AttrSchema` join lattice; serde DTO JSON shape; sampler accumulator edge cases (Maybe presence, type-conflict→Unknown, first-seen ordering); each transfer function (incl. conservative-presence and dynamic-port cases).
- Integration: `schema_sample` against a real GeoJSON read (tempfile); the `probe-schema` command end-to-end on `GeoJsonReader → AttributeManager(remove)` — asserts the reader exposes real keys (closed schema) and the removed key is absent downstream while others survive; OutputRouter `routingPort` and FeatureFilter condition ports appear in the inferred output.
- Manual e2e on the real PLATEAU `quality-check/bldg` workflow (87 nodes, `!include`s): emits valid JSON, no panic.
- Full local CI parity: `cargo make check`, `format --check`, `clippy -D warnings`, `format-taplo --check`, `check-schema` (0 drift), `check-generate-examples-cms-workflow` (0 drift), `test-rs` — all green. Engine version `0.0.380`.

## Screenshot

```
$ reearth-flow probe-schema --workflow demo.yml
"GeoJsonReader":     { default: [ myAttribute:String(always), photoURL:String(always), address:String(maybe) ] }
"RemoveMyAttribute": { default: [ photoURL:String(always), address:String(maybe) ] }   # myAttribute removed
```

## Which point I want you to review particularly

- **Sampling mechanism** (`schema_sample.rs`): running a source reader on a current-thread tokio runtime over a bounded mpsc channel, then dropping the receiver to stop it; features are folded into an accumulator and discarded as they arrive. No-panic + no-deadlock + bounded memory are the key invariants.
- **`open` + `note` fallback semantics**: sources without a resolvable dataset (or expression-driven sources) report `open: true` + a note rather than failing — so the editor flow degrades gracefully.
- **Inference fidelity vs. runtime**: presence is intentionally conservative (no key claimed `Always` the runtime might not produce), and dynamic router/filter ports are derived from `with` to match runtime routing.

## Memo

Engine foundation only. `probe-schema` is additive and does not affect `run`/`dot`. Copilot review comments (presence over-claim, dynamic ports, sampler memory) have been addressed and the threads resolved. Branch based on `origin/main`.

### CODE REVIEW

File: engine/runtime/action-processor/src/feature/filter.rs [hunk 1]
```diff
@@ -105,6 +105,29 @@ impl ProcessorFactory for FeatureFilterFactory {
...
+            .cloned()
+            .unwrap_or_else(AttrSchema::open);
+
+        let map = self
+            .get_output_ports()
+            .into_iter()
+            .map(|port| (port, input.clone()))
+            .collect();
+        Some(map)
+    }
```
[the remark this answers is not in the document]
Comment (pyshx) [reply engine/runtime/action-processor/src/feature/filter.rs:141]: Fixed in b056aa4d: `FeatureFilter::infer_output_schema` now maps the input schema onto the dynamic `conditions[].outputPort` ports in addition to the static `unfiltered` port (identity passthrough, since the filter routes whole features unchanged).

File: engine/runtime/runtime/src/schema_infer.rs [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+                None => {
+                    // Schema-transparent fallback: pass the join of all inputs
+                    // through to every declared output port.
+                    let joined = join_all_inputs(&inputs);
+                    factory
+                        .output_ports()
+                        .into_iter()
+                        .map(|p| (p, joined.clone()))
+                        .collect()
+                }
```
[the remark this answers is not in the document]
Comment (pyshx) [reply engine/runtime/runtime/src/schema_infer.rs:126]: Fixed in b056aa4d: added `effective_output_ports` which augments `get_output_ports()` with dynamic ports parsed from `with` (OutputRouter `routingPort`, FeatureFilter `conditions[].outputPort`), mirroring `builder_dag`. The passthrough fallback now uses it, so OutputRouter's routed port appears in the report.

File: engine/runtime/runtime/src/schema_infer.rs [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+                match factory.infer_output_schema(&inputs, &node.with) {
+                    Some(map) => map,
+                    None => {
+                        let joined = join_all_inputs(&inputs);
+                        factory
+                            .output_ports()
+                            .into_iter()
+                            .map(|p| (p, joined.clone()))
+                            .collect()
+                    }
```
[the remark this answers is not in the document]
Comment (pyshx) [reply engine/runtime/runtime/src/schema_infer.rs:212]: Fixed in b056aa4d: same `effective_output_ports` helper is now used in the `infer_with_sampling` passthrough fallback too, so OutputRouter's `routingPort` is included.

File: engine/runtime/action-processor/src/attribute/manager.rs [hunk 4]
```diff
@@ -72,6 +72,53 @@ impl ProcessorFactory for AttributeManagerFactory {
...
+                // Rename's destination name is an expression -> not statically knowable.
+                // Drop the source key and mark the schema open (an unknown-named attr appears).
+                Method::Rename => {
+                    out.fields.shift_remove(&attr);
+                    out.open = true;
+                }
+                Method::Remove => {
+                    out.fields.shift_remove(&attr);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (pyshx) [reply engine/runtime/action-processor/src/attribute/manager.rs:128]: Fixed in 6010cb40: inference is now conservative. Create/Convert insert a newly-produced key as `Maybe` (runtime only inserts on successful eval; Convert also skips when the source is absent); a key already present keeps its presence and just gets type `Unknown`. Rename no longer unconditionally drops the source — it downgrades the source key to `Maybe` (runtime keeps it on eval error / destination collision) and sets `open` for the dynamic destination name.

File: engine/runtime/action-processor/src/attribute/manager.rs [hunk 5]
```diff
@@ -271,3 +318,109 @@ fn convert_single_operation(operations: &[Operation]) -> super::errors::Result<V
...
+            .expect("default port present");
+
+        assert_eq!(
+            schema.fields.get(&Attribute::new("bar".to_string())),
+            Some(&AttrField::always(AttrType::String))
+        );
+        assert_eq!(
+            schema.fields.get(&Attribute::new("foo".to_string())),
+            Some(&AttrField::always(AttrType::Unknown))
+        );
```
[the remark this answers is not in the document]
Comment (pyshx) [reply engine/runtime/action-processor/src/attribute/manager.rs:385]: Fixed in 6010cb40: the create test now expects `Maybe` presence for the newly-created key.

File: engine/runtime/action-processor/src/attribute/manager.rs [hunk 6]
```diff
@@ -271,3 +318,109 @@ fn convert_single_operation(operations: &[Operation]) -> super::errors::Result<V
...
+        let out = AttributeManagerFactory
+            .infer_output_schema(&inputs, &with)
+            .expect("inference should succeed");
+        let schema = out
+            .get(&DEFAULT_PORT.clone())
+            .expect("default port present");
+
+        assert!(!schema.fields.contains_key(&Attribute::new("a".to_string())));
+        assert!(schema.open);
+    }
```
[the remark this answers is not in the document]
Comment (pyshx) [reply engine/runtime/action-processor/src/attribute/manager.rs]: Fixed in 6010cb40: the rename test (renamed to `infer_rename_downgrades_source_and_sets_open`) now expects the source key to remain with downgraded `Maybe` presence while the schema is marked `open`.

File: engine/runtime/runtime/src/schema_sample.rs [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+}
+
+/// Spawn the source's `start`, drain features off the channel up to `sample_size`,
+/// then drop the receiver (so a still-running source stops on a closed channel)
+/// and join the task. Errors from the source are surfaced as a `String`.
+async fn read_features(
+    mut source: Box<dyn crate::node::Source>,
+    ctx: NodeContext,
+    sample_size: usize,
+) -> Result<Vec<Feature>, String> {
```
[the remark this answers is not in the document]
Comment (pyshx) [reply engine/runtime/runtime/src/schema_sample.rs]: Fixed in 355d658e: the sampler now folds each feature into a `SchemaAccumulator` (per-attribute type + count) as it arrives and drops the feature immediately, so memory is O(#attributes) instead of O(#features) — including the `--sample-size 0` unbounded case. Schema output is identical (verified by the existing union tests).

**Review by n4to4 — APPROVED** [review]
👍
